pyo3-stub-gen/src/util.rs

```diff
@@ -0,0 +1,145 @@
+use pyo3::{prelude::*, types::*};
+
+/// Reference to <https://github.com/Jij-Inc/serde-pyobject/blob/5916fefc5b7f0e096ed13869dc41b36ad6f62dc2/src/de.rs#L309-L333>
+#[allow(clippy::result_unit_err)]
+pub fn fmt_py_obj(any: &Bound<'_, PyAny>) -> Result<String, ()> {
+    if any.is_instance_of::<PyDict>() {
+        return any.downcast::<PyDict>().map_err(|_| ()).and_then(|dict| {
+            dict.into_iter()
+                .map(|(k, v)| {
+                    if let (Ok(k), Ok(v)) = (fmt_py_obj(&k), fmt_py_obj(&v)) {
+                        Ok(format!("{k}: {v}"))
+                    } else {
+                        Err(())
+                    }
+                })
+                .collect::<Result<Vec<String>, ()>>()
+                .map(|kv_list| format!("{{{}}}", kv_list.join(", ")))
+        });
+    }
+    if any.is_instance_of::<PyList>() {
+        return any.downcast::<PyList>().map_err(|_| ()).and_then(|list| {
+            list.into_iter()
+                .map(|v| fmt_py_obj(&v))
+                .collect::<Result<Vec<String>, ()>>()
+                .map(|v_list| format!("[{}]", v_list.join(", ")))
+        });
+    }
+    if any.is_instance_of::<PyTuple>() {
+        return any.downcast::<PyTuple>().map_err(|_| ()).and_then(|list| {
+            list.into_iter()
+                .map(|v| fmt_py_obj(&v))
+                .collect::<Result<Vec<String>, ()>>()
+                .map(|v_list| format!("({})", v_list.join(", ")))
```

Comment:
This is one manifestation of corner-cases of manual formatting. I think it formats the tuple `(42,)`  (i.e. a singleton tuple with `42` its only component) as `(42)`, which will be interpreted as a mere number literal `42`.